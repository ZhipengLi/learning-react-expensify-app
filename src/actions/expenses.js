import uuid from 'uuid';
import database from '../firebase/firebase';


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expensesData = {}) => {
    return (dispatch, getState) => {
      const uid = getState().auth.uid;
      const {
          description = '', 
          note = '', 
          amount = 0, 
          createdAt = 0
      } = expensesData;
      const expense = {description, note, amount, createdAt };
      console.log(expense);
      return database.ref(`users/${uid}/expenses`).push(expense).then(
          (ref)=> {
              //console.log('after expense is pushed to store', ref.key);
              dispatch(addExpense({
                  id: ref.key,
                  ...expense
              }))
              console.log('after expense is pushed to store', ref.key);
          }
      ).catch((e)=>{
          console.log('database error', e);
      });
    };
};
// REMOVE_EXPENSE
export const removeExpense = (
    { 
        id 
    } = {}) => (
    {
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({id}) => {
    return (dispatch, getState) => {
      const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(
            ()=> {
                dispatch(removeExpense({
                    id
                }))
            }
        ).catch((e)=>{
            console.log('database error', e);
        });
    };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) =>({
    type: 'EDIT_EXPENSE',
    id,
    updates
});
export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
      const uid = getState().auth.uid;
      return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(
          ()=> {
              //console.log('after expense is pushed to store', ref.key);
              dispatch(editExpense(id, updates));
          }
      ).catch((e)=>{
          console.log('database error', e);
      });
    };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//export const startSetExpenses;
export const startSetExpenses = () => {
    return (dispatch, getState) => {
      const uid = getState().auth.uid;
      return database.ref(`users/${uid}/expenses`).once('value').then(
          (snapshot)=> {
//                console.log('whats is in expenses:', snapshot.val());
              const expenses = [];
              snapshot.forEach(element => {
                  expenses.push({
                      id: element.key,
                      ...element.val()
                  });
              });
              dispatch(setExpenses(expenses));
              console.log('after expense is set to store',expenses);
          }
      ).catch((e)=>{
          console.log('database error', e);
      });
    };
};

