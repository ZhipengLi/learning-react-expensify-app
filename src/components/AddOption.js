import React from 'react';

export default class AddOption extends React.Component {
    state = {
        error:undefined
    }

    handleOnSubmit = (e) => {
        e.preventDefault();
        const option  = e.target.elements.newOption.value;
        const error = this.props.handleAddOption(option);
        this.setState(() => {
            return {
                error: error
            }
        });   
        if (!error) {
            e.target.elements.newOption.value = '';
        }    
    }
    render() {
        return (
            <div>
                {this.state.error && <p className='add-option-error'>{this.state.error}</p>}
                <form className='add-option' onSubmit={this.handleOnSubmit}>
                    <input className='add-option__input' name="newOption" />
                    <button className='button'>Add option</button>
                </form>
            </div>
        );
    }
}