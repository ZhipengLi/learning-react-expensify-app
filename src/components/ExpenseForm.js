import React from 'react';
import moment, { calendarFormat } from 'moment';
import { SingleDatePicker } from 'react-dates';

//const date = new Date();
//const now = moment();
//console.log(now.format('MMM'));

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
        //console.log(props.expense);
        //console.log(props.expense.description, props.expense.amount);
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        console.log(amount);
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?/)) {
            this.setState(() => ({amount}));
        }
    }
    onDateChange = ( createdAt ) => {
        if( createdAt ) {
            this.setState( () => ({createdAt }));
        }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
      };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            // Set error state equal to 'Please provide description an amount';
            this.setState( ()=>({error: 'Please provide description or amount'}));
            //console.log('inside error', this.state.description, this.state,amount);
        }
        else {
            console.log(this.state);
            this.setState(()=> ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }
    render() {
        return (
          <form className='form' onSubmit={this.onSubmit}>
            {this.state.error && <p className='form__error'>{this.state.error}</p> }
              <input 
                  type='text'
                  placeholder='Description'
                  autoFocus
                  className='text-input'
                  value={this.state.description}
                  onChange={this.onDescriptionChange}
              />
              <input 
                  type='number'
                  placeholder='amount'
                  className='text-input'
                  value={this.state.amount}
                  onChange={this.onAmountChange}
              />
              <SingleDatePicker 
                  date={this.state.createdAt}
                  onDateChange={this.onDateChange}
                  focused={this.state.calendarFocused}
                  onFocusChange={this.onFocusChange}
                  numberOfMonths={1}
                  isOutsideRange={(day) => false}
              />
              <textarea placeholder='Add a note for your expense.'
                className='textarea'
                value={this.state.note}
                onChange={this.onNoteChange}
              >
              </textarea>
              <div>
                <button>Save Expense</button>
              </div>
          </form>
        );
    }
}
