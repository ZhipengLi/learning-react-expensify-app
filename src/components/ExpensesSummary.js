import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import SelectExpenses from '../selectors/expenses';
import SelectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export class ExpensesSummary extends React.Component{
    render(){
        const expenseWord = this.props.expensesCount === 1 ? 'expense' : 'expenses';
        return (<div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>
              Viewing <span>{ this.props.expensesCount }</span> {expenseWord} totaling <span>{ numeral(this.props.expensesTotal/100).format('$0, 0.00')}</span>
            </h1>  
            <div className='page-header__actions'>
              <Link className='button' to='/create'>Add Expense</Link>
            </div>
          </div>
        </div>);
    }
}

const mapStateToProps = (state, props) => {
    const visibleExpenses = SelectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses ? visibleExpenses.length: 0,
        expensesTotal: SelectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);