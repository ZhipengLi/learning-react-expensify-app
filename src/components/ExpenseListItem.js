
import React from 'react';
import {Link} from 'react-router-dom';
import { connect } from 'react-redux';
//import { ProgressPlugin } from 'webpack';
import { removeExpense } from '../actions/expenses';

export const ExpenseListItem = ({dispatch, id, description, amount, createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <h1>{description}</h1>
        <p>{amount} -- {createdAt}</p>
    </div>
);
//const stateToProps = (state) => {
//    return {};
//}
export default connect()(ExpenseListItem);
//export default ExpenseListItem;