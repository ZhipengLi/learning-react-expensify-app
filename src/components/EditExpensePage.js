import React from 'react';
import { connect} from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
    //constructor(props) {
    //    super();
    //    if(props.expense) {
    //        this.state= {expense: props.expense};
    //    }
    //}
    onSubmit = (expense) => {
        //console.log('onsubmit is called in EditExpensePage', expense);
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }
    onRemove = () => {
        //console.log('state expense', this.state.expense);
        this.props.removeExpense({id: this.props.expense.id});
        this.props.history.push('/');
    }
    render() {
        return (
          <div>
            <div className='page-header'>
              <div className='content-container'>
                <h1 className='page-header__title'>Edit Expense</h1>
              </div>
            </div>
            <div className='content-container'>
              <ExpenseForm 
              expense={this.props.expense}
              onSubmit={(expense) => {this.onSubmit(expense);}}
              />
              <button className='button button--secondary' onClick={() => {this.onRemove()}}>
              Remove Expense
              </button>  
            </div>
          </div>
        );
    }
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find( (exp) => {
            //console.log(`${exp.id} === ${props.match.params.id}`);
            return exp.id === props.match.params.id;
        })
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    removeExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);