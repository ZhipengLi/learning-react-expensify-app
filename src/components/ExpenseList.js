import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import SelectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className='content-container'>
      <div className='list-header'>
        <div className='show-for-mobile'>Expenses</div>
        <div className='show-for-desktop'>Expense</div>
        <div className='show-for-desktop'>Amount</div>
      </div>
      <div className='list-body'>
        {
            props.expenses.length === 0 ? (
              <div className='list-item list-item--message'>
               <p>No expenses</p>
              </div>
            ) : (
                props.expenses.map( (exp) => {
                    return <div key = {exp.id}><ExpenseListItem {...exp} /></div>;
                }) 
            )
        }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: SelectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);