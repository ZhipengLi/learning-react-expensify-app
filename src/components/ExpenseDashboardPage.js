import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

export const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
        <ExpensesSummary />
    </div>
);

export default ExpenseDashboardPage;