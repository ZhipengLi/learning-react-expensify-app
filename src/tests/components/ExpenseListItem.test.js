import React from 'react';
import { shallow} from 'enzyme';
import { ExpenseListItem } from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test(`should return expense list item`, () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
});

//{dispatch, id, description, amount, createdAt}