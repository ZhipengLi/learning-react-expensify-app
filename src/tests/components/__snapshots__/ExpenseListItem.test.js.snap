// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return expense list item 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <h1>
    Gum
  </h1>
  <p>
    $1.95
    -
    December 31st, 1969
  </p>
</div>
`;
